File: intellij-plugin/src/main/java/com/kibernet/luaunity/generation/LuaAnnotationGenerator.java
package com.kibernet.luaunity.generation;

import com.kibernet.luaunity.model.LuaUnityClass;
import com.kibernet.luaunity.model.LuaUnityMember;
import com.kibernet.luaunity.model.LuaUnityMethod;
import com.kibernet.luaunity.model.LuaUnityParameter;
import com.kibernet.luaunity.model.LuaUnityTypeRef;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.TreeSet;
import java.util.stream.Stream;

public final class LuaAnnotationGenerator {
    private static final Map<String, String> TYPE_MAP = Map.ofEntries(
            Map.entry("System.Single", "number"),
            Map.entry("System.Double", "number"),
            Map.entry("System.Decimal", "number"),
            Map.entry("System.Int16", "number"),
            Map.entry("System.Int32", "number"),
            Map.entry("System.Int64", "number"),
            Map.entry("System.SByte", "number"),
            Map.entry("System.UInt16", "number"),
            Map.entry("System.UInt32", "number"),
            Map.entry("System.UInt64", "number"),
            Map.entry("System.Boolean", "boolean"),
            Map.entry("System.String", "string"),
            Map.entry("System.Object", "table"),
            Map.entry("System.Void", "void")
    );

    public int generate(List<LuaUnityClass> classes, Path outputDirectory) throws IOException {
        deleteDirectory(outputDirectory);
        Files.createDirectories(outputDirectory);

        Set<String> namespaces = collectNamespaces(classes);
        for (String namespace : namespaces) {
            writeNamespace(outputDirectory, namespace);
        }
        for (LuaUnityClass luaClass : classes) {
            writeClass(outputDirectory, luaClass);
        }

        return namespaces.size() + classes.size();
    }

    private void writeNamespace(Path outputDirectory, String namespace) throws IOException {
        String[] parts = namespace.split("\\.");
        List<String> lines = new ArrayList<>();
        for (int i = 1; i <= parts.length; i++) {
            String current = String.join(".", List.of(parts).subList(0, i));
            lines.add(current + " = " + current + " or {}");
        }
        Files.writeString(outputDirectory.resolve(namespace + ".ns.lua"), String.join("\n", lines) + "\n", StandardCharsets.UTF_8);
    }

    private void writeClass(Path outputDirectory, LuaUnityClass luaClass) throws IOException {
        List<String> lines = new ArrayList<>();
        for (LuaUnityMember field : luaClass.getFields()) {
            lines.add("---@field public " + sanitizeIdentifier(field.getName()) + " " + renderType(field.getType()));
        }
        for (LuaUnityMember property : luaClass.getProperties()) {
            lines.add("---@field public " + sanitizeIdentifier(property.getName()) + " " + renderType(property.getType()));
        }

        if (luaClass.getBaseType() == null || luaClass.getBaseType().isBlank()) {
            lines.add("---@class " + luaClass.getFullName());
        } else {
            lines.add("---@class " + luaClass.getFullName() + " : " + convertTypeName(luaClass.getBaseType()));
        }
        lines.add("local m = {}");

        for (List<LuaUnityMethod> methods : groupByName(luaClass.getMethods()).values()) {
            LuaUnityMethod first = methods.get(0);
            lines.add("");
            for (int i = 1; i < methods.size(); i++) {
                lines.add(renderOverload(methods.get(i)));
            }
            for (LuaUnityParameter parameter : first.getParameters()) {
                lines.add("---@param " + sanitizeIdentifier(parameter.getName()) + " " + renderType(parameter.getType()));
            }
            String returnType = renderType(first.getReturnType());
            if (!"void".equals(returnType)) {
                lines.add("---@return " + returnType);
            }
            String receiver = first.isStaticMethod() ? "." : ":";
            String parameters = first.getParameters().stream()
                    .map(parameter -> sanitizeIdentifier(parameter.getName()))
                    .reduce((left, right) -> left + ", " + right)
                    .orElse("");
            lines.add("function m" + receiver + sanitizeIdentifier(first.getName()) + "(" + parameters + ") end");
        }

        lines.add(luaClass.getFullName() + " = m");
        lines.add("return m");
        Files.writeString(outputDirectory.resolve(luaClass.getFullName() + ".lua"), String.join("\n", lines) + "\n", StandardCharsets.UTF_8);
    }

    private Set<String> collectNamespaces(List<LuaUnityClass> classes) {
        Set<String> namespaces = new TreeSet<>();
        for (LuaUnityClass luaClass : classes) {
            String[] parts = luaClass.getFullName().split("\\.");
            for (int i = 1; i < parts.length; i++) {
                namespaces.add(String.join(".", List.of(parts).subList(0, i)));
            }
        }
        return namespaces;
    }

    private Map<String, List<LuaUnityMethod>> groupByName(List<LuaUnityMethod> methods) {
        Map<String, List<LuaUnityMethod>> grouped = new LinkedHashMap<>();
        for (LuaUnityMethod method : methods) {
            grouped.computeIfAbsent(method.getName(), ignored -> new ArrayList<>()).add(method);
        }
        return grouped;
    }

    private String renderOverload(LuaUnityMethod method) {
        String parameters = method.getParameters().stream()
                .map(parameter -> sanitizeIdentifier(parameter.getName()) + ": " + renderType(parameter.getType()))
                .reduce((left, right) -> left + ", " + right)
                .orElse("");
        String returnType = renderType(method.getReturnType());
        return "void".equals(returnType)
                ? "---@overload fun(" + parameters + ")"
                : "---@overload fun(" + parameters + "): " + returnType;
    }

    private String renderType(LuaUnityTypeRef typeRef) {
        String converted = convertTypeName(typeRef.getName());
        return typeRef.isArray() ? converted + "[]" : converted;
    }

    private String convertTypeName(String name) {
        return TYPE_MAP.getOrDefault(name, name);
    }

    private String sanitizeIdentifier(String name) {
        String safe = name.replaceAll("[^A-Za-z0-9_]", "_");
        return safe.matches("^[A-Za-z_].*") ? safe : "_" + safe;
    }

    private void deleteDirectory(Path directory) throws IOException {
        if (!Files.exists(directory)) {
            return;
        }
        try (Stream<Path> walk = Files.walk(directory)) {
            List<Path> paths = walk.sorted(Comparator.reverseOrder()).toList();
            for (Path path : paths) {
                Files.deleteIfExists(path);
            }
        }
    }
}

File: vscode-plugin/src/LuaAnnotationGenerator.ts
import * as fs from 'fs/promises';
import * as path from 'path';

export interface LuaUnityTypeRef {
    name: string;
    isArray?: boolean;
}

export interface LuaUnityMember {
    name: string;
    type: LuaUnityTypeRef;
}

export interface LuaUnityMethod {
    name: string;
    isStatic: boolean;
    parameters: LuaUnityParameter[];
    returnType: LuaUnityTypeRef;
}

export interface LuaUnityParameter {
    name: string;
    type: LuaUnityTypeRef;
}

export interface LuaUnityClass {
    fullName: string;
    baseType?: string;
    fields: LuaUnityMember[];
    properties: LuaUnityMember[];
    methods: LuaUnityMethod[];
}

const typeMap = new Map<string, string>([
    ['System.Single', 'number'],
    ['System.Double', 'number'],
    ['System.Decimal', 'number'],
    ['System.Int16', 'number'],
    ['System.Int32', 'number'],
    ['System.Int64', 'number'],
    ['System.SByte', 'number'],
    ['System.UInt16', 'number'],
    ['System.UInt32', 'number'],
    ['System.UInt64', 'number'],
    ['System.Boolean', 'boolean'],
    ['System.String', 'string'],
    ['System.Object', 'table'],
    ['System.Void', 'void']
]);

export class LuaAnnotationGenerator {
    async generate(classes: LuaUnityClass[], outputDirectory: string): Promise<number> {
        await fs.rm(outputDirectory, { recursive: true, force: true });
        await fs.mkdir(outputDirectory, { recursive: true });

        const namespaces = collectNamespaces(classes);
        for (const namespace of namespaces) {
            await this.writeNamespace(outputDirectory, namespace);
        }

        for (const clazz of classes) {
            await this.writeClass(outputDirectory, clazz);
        }

        return namespaces.length + classes.length;
    }

    private async writeNamespace(outputDirectory: string, namespace: string): Promise<void> {
        const filePath = path.join(outputDirectory, `${namespace}.ns.lua`);
        const lines: string[] = [];
        const parts = namespace.split('.');

        for (let i = 0; i < parts.length; i++) {
            const current = parts.slice(0, i + 1).join('.');
            lines.push(`${current} = ${current} or {}`);
        }

        await fs.writeFile(filePath, `${lines.join('\n')}\n`, 'utf8');
    }

    private async writeClass(outputDirectory: string, clazz: LuaUnityClass): Promise<void> {
        const filePath = path.join(outputDirectory, `${clazz.fullName}.lua`);
        const lines: string[] = [];

        for (const field of clazz.fields) {
            lines.push(`---@field public ${sanitizeIdentifier(field.name)} ${renderType(field.type)}`);
        }
        for (const property of clazz.properties) {
            lines.push(`---@field public ${sanitizeIdentifier(property.name)} ${renderType(property.type)}`);
        }

        const classDeclaration = clazz.baseType
            ? `---@class ${clazz.fullName} : ${convertTypeName(clazz.baseType)}`
            : `---@class ${clazz.fullName}`;
        lines.push(classDeclaration);
        lines.push('local m = {}');

        const groupedMethods = groupByName(clazz.methods);
        for (const methods of groupedMethods.values()) {
            const [first, ...overloads] = methods;
            lines.push('');
            for (const overload of overloads) {
                lines.push(renderOverload(overload));
            }
            for (const parameter of first.parameters) {
                lines.push(`---@param ${sanitizeIdentifier(parameter.name)} ${renderType(parameter.type)}`);
            }
            if (renderType(first.returnType) !== 'void') {
                lines.push(`---@return ${renderType(first.returnType)}`);
            }

            const receiver = first.isStatic ? '.' : ':';
            const parameters = first.parameters.map(parameter => sanitizeIdentifier(parameter.name)).join(', ');
            lines.push(`function m${receiver}${sanitizeIdentifier(first.name)}(${parameters}) end`);
        }

        lines.push(`${clazz.fullName} = m`);
        lines.push('return m');

        await fs.writeFile(filePath, `${lines.join('\n')}\n`, 'utf8');
    }
}

function collectNamespaces(classes: LuaUnityClass[]): string[] {
    const namespaces = new Set<string>();
    for (const clazz of classes) {
        const parts = clazz.fullName.split('.');
        for (let i = 1; i < parts.length; i++) {
            namespaces.add(parts.slice(0, i).join('.'));
        }
    }
    return [...namespaces].sort((a, b) => a.localeCompare(b));
}

function groupByName(methods: LuaUnityMethod[]): Map<string, LuaUnityMethod[]> {
    const grouped = new Map<string, LuaUnityMethod[]>();
    for (const method of methods) {
        if (!grouped.has(method.name)) {
            grouped.set(method.name, []);
        }
        grouped.get(method.name)!.push(method);
    }
    return grouped;
}

function renderOverload(method: LuaUnityMethod): string {
    const params = method.parameters
        .map(parameter => `${sanitizeIdentifier(parameter.name)}: ${renderType(parameter.type)}`)
        .join(', ');
    const returnType = renderType(method.returnType);
    return returnType === 'void'
        ? `---@overload fun(${params})`
        : `---@overload fun(${params}): ${returnType}`;
}

function renderType(type: LuaUnityTypeRef): string {
    const converted = convertTypeName(type.name);
    return type.isArray ? `${converted}[]` : converted;
}

function convertTypeName(name: string): string {
    return typeMap.get(name) ?? name;
}

function sanitizeIdentifier(name: string): string {
    const safe = name.replace(/[^A-Za-z0-9_]/g, '_');
    return /^[A-Za-z_]/.test(safe) ? safe : `_${safe}`;
}
